File: 020 Valid Parentheses/valid_parentheses.cs
/*	
 *	LeetCode Submissions by Xinyu Chen
 *	Valid Parentheses
 *	https://leetcode.com/problems/valid-parentheses/
 *	Runtime: 163 ms
 */

public class Solution {
    public bool IsValid(string s) {
    	Dictionary<char, char> pairs = new Dictionary<char, char>();
        pairs.Add(')', '(');
		pairs.Add('}', '{');
		pairs.Add(']', '[');
		Stack<char> st = new Stack<char>();
		foreach (var c in s)
		{
			if (c == '(' || c == '{' || c == '[') st.Push(c);
			else if (st.Count == 0) return false;
			else if (pairs[c] != st.Pop()) return false;
		}
		if (st.Count == 0) return true;
		return false;
    }
}

File: 020 Valid Parentheses/valid_parentheses.py
# LeetCode Submissions by Xinyu Chen
# Valid Parentheses
# https://leetcode.com/problems/valid-parentheses/
# Runtime: 46 ms

class Solution:
    # @param {string} s
    # @return {boolean}
    def isValid(self, s):
        st = []
        pairs = {')':'(', '}':'{', ']':'['}
        for c in s:
            if c == '(' or c == '{' or c == '[':
                st.append(c)
            elif len(st) == 0:
                return False
            else:
                if pairs[c] != st.pop():
                    return False
        if len(st) == 0:
            return True
        return False